edx_event_bus_redis/internal/producer.py

```diff
@@ -0,0 +1,135 @@
+"""
+Produce Redis events from signals.
+
+Main function is ``create_producer()``, which should be referred to from ``EVENT_BUS_PRODUCER``.
+"""
+
+import logging
+from typing import Optional
+
+import attr
+from edx_django_utils.monitoring import record_exception
+from openedx_events.data import EventsMetadata
+from openedx_events.event_bus import EventBusProducer
+from openedx_events.tooling import OpenEdxPublicSignal
+
+from .config import get_full_topic, load_common_settings
+from .utils import AUDIT_LOGGING_ENABLED
+
+logger = logging.getLogger(__name__)
+
+
+def record_producing_error(error, context):
+    """
+    Record an error in producing an event to both the monitoring system and the regular logs
+
+    Arguments:
+        error: The exception or error raised during producing
+        context: An instance of ProducingContext containing additional information about the message
+    """
+    try:
+        # record_exception() is a wrapper around a New Relic method that can only be called within an except block,
+        # so first re-raise the error
+        raise Exception(error)  # pylint: disable=broad-exception-raised
+    except BaseException:
+        record_exception()
+        logger.exception(f"Error delivering message to Redis event bus. {error=!s} {context!r}")
+
+
+@attr.s(kw_only=True, repr=False)
+class ProducingContext:
+    """
+    Wrapper class to allow us to link a call to produce() with the on_event_deliver callback
+    """
+    full_topic = attr.ib(type=str, default=None)
+    event_key = attr.ib(type=str, default=None)
+    signal = attr.ib(type=OpenEdxPublicSignal, default=None)
+    initial_topic = attr.ib(type=str, default=None)
+    event_key_field = attr.ib(type=str, default=None)
+    event_data = attr.ib(type=dict, default=None)
+    event_metadata = attr.ib(type=EventsMetadata, default=None)
+
+    def __repr__(self):
+        """Create a logging-friendly string"""
+        return " ".join([f"{key}={value!r}" for key, value in attr.asdict(self, recurse=False).items()])
+
+    def on_event_deliver(self, err, evt):
+        """
+        Simple callback method for debugging event production
+
+        If there is any error, log all the known information about the calling context so the event can be recreated
+        and/or resent later. This log will not contain the exact headers but will contain the EventsMetadata object
+        that can be used to recreate them.
+
+        Arguments:
+            err: Error if event production failed
+            evt: Event that was delivered (or failed to be delivered)
+        """
+        if err is not None:
+            record_producing_error(err, self)
+        else:
+            # Audit logging on success. See ADR: docs/decisions/0010_audit_logging.rst
```

Comment:
We should expand this link to go back to the Kafka docs, I think. In fact, as part of this work we should move any ADRs from the Kafka repo to openedx-events as more generally useful.